src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,487 @@
+    case Attributes::OnMode::Id:
+        if (mOnMode.IsNull())
+        {
+            ReturnErrorOnFailure(aEncoder.EncodeNull());
+        }
+        else
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(mOnMode));
+        }
+        break;
+    case Attributes::SupportedModes::Id:
+        if (NumberOfModes() == 0)
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+
+        Instance * d   = this;
+        CHIP_ERROR err = aEncoder.EncodeList([d](const auto & encoder) -> CHIP_ERROR {
+            for (uint8_t i = 0; i < d->NumberOfModes(); i++)
+            {
+                ModeOptionStructType mode;
+                ChipError err1 = CHIP_NO_ERROR;
+
+                // Get the mode label
+                char buffer[64];
+                MutableCharSpan label(buffer);
+                err1 = d->GetModeLabelByIndex(i, label);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.label = label;
+
+                // Get the mode value
+                err1 = d->GetModeValueByIndex(i, mode.mode);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+
+                // Get the mode tags
+                ModeTagStructType tagsBuffer[8];
+                DataModel::List<ModeTagStructType> tags(tagsBuffer);
+                err1 = d->GetModeTagsByIndex(i, tags);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.modeTags = tags;
+
+                ReturnErrorOnFailure(encoder.Encode(mode));
+            }
+            return CHIP_NO_ERROR;
+        });
+        ReturnErrorOnFailure(err);
+        break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+// Implements checking before attribute writes.
+CHIP_ERROR Instance::Write(const ConcreteDataAttributePath & attributePath, AttributeValueDecoder & aDecoder)
+{
+    DataModel::Nullable<uint8_t> newMode;
+    ReturnErrorOnFailure(aDecoder.Decode(newMode));
+    Status status;
+
+    if (!newMode.IsNull()) // This indicates that the new mode is null.
+    {
+        if (!IsSupportedMode(newMode.Value()))
+        {
+            return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+        }
+    }
+
+    switch (attributePath.mAttributeId)
+    {
+    case ModeBase::Attributes::StartUpMode::Id:
+        status = UpdateStartUpMode(newMode);
+        return StatusIB(status).ToChipError();
+    case ModeBase::Attributes::OnMode::Id:
+        status = UpdateOnMode(newMode);
+        return StatusIB(status).ToChipError();
+    }
+
+    return StatusIB(Protocols::InteractionModel::Status::InvalidCommand).ToChipError();
+}
+
+Status Instance::UpdateCurrentMode(uint8_t aNewMode)
+{
+    if (!IsSupportedMode(aNewMode))
+    {
+        return Protocols::InteractionModel::Status::InvalidValue;
```

Comment:
Please don't use deprecated status values.

That said, if we are checking this in here, why are we also checking it in Instance::Write?  Just seems like extra code to check both places....  I would guess we should keep the checks here, since these can be called by the application, and remove them in Write().  And use ConstraintError here, not InvalidValue.

Reply:
How can I find out what is depricated and what is supposed to be used instead of it?

Reply:
I replaced `InvalidValue` with `ConstraintError`.

Reply:
> How can I find out what is depricated and what is supposed to be used instead of it?

@hicklin definition of Invalid value says:

```
src/protocols/interaction_model/StatusCode.h:    InvalidValue = ConstraintError, // Deprecated
```